# Matrix Operations

In [2]:
# import libraries
import pandas as pd
import numpy as np

**What are Matrices?**

A matrix is a two-dimensional array of numbers arranged in rows and columns.

Matrices are used to represent datasets, where each row corresponds to an observation or sample, and each column represents a feature or attribute of that sample.

In this project, I'll be using a movie dataset where each row represents a movie, and each column represents a feature.


In [3]:
# load data into dataframe
df = pd.read_csv('../datasets/Hydra-Movie-Scrape.csv')


In [4]:
# create movie feature matrix
movie_matrix = df[['Runtime', 'Rating']].to_numpy()

Now that we have our feature matrix, let's check its shape

In [5]:
movie_matrix.shape

(3886, 2)

It is a 3886 x 2 matrix, meaning it has 3886 rows and 2 columns. Since the number of rows is much larger than the number of columns, this is a tall matrix.

In [6]:
# extract movie titles
movie_titles = df['Title']


Let us check out the rows and columns.

In [15]:
# calculate mean of rows
np.mean(movie_matrix[:5], axis=1) # first 5 rows

array([36.7 , 41.45, 35.9 ,  3.3 , 53.3 ])

In [8]:
# calculate mean of columns
np.mean(movie_matrix, axis=0)

array([100.43952651,   6.55892949])

The mean runtime for this dataset is 100.44 minutes and mean rating is 6.56

In [9]:
# find max runtime
runtime_col = movie_matrix[:, 0]
max_run = float(np.max(runtime_col))
id_run = np.argmax(runtime_col)

print(f"The movie with the longest runtime is \"{movie_titles[id_run]}\" with a runtime of {round(max_run/60, 2)} hours.")

The movie with the longest runtime is "From What Is Before" with a runtime of 5.63 hours.


A 5 hour 6 minutes long movie? Damn, I wonder what kind of story needs that much time to tell.

In [14]:
# find max rating
rating_col = movie_matrix[:, 1]
max_rate = float(np.max(rating_col))
id_rate = np.argmax(rating_col)

print(f"The movie with the highest rating is \"{movie_titles[id_rate]}\" with a rating of {max_rate}")

The movie with the highest rating is "The Mountain II" with a rating of 9.5


Now let's check out the Transpose of the matrix.

In [11]:
movie_transpose = movie_matrix.T

In [18]:
print(f"First five rows of original matrix: \n{movie_matrix[:5]}\n")

print(f"First five columns of transposed matrix: \n{movie_transpose[:, :5]}")

First five rows of original matrix: 
[[66.   7.4]
 [75.   7.9]
 [65.   6.8]
 [ 0.   6.6]
 [99.   7.6]]

First five columns of transposed matrix: 
[[66.  75.  65.   0.  99. ]
 [ 7.4  7.9  6.8  6.6  7.6]]


In this notebook, the movie dataset was represented as a matrix where rows correspond to movies and columns represent numerical features. Viewing the data in matrix form makes it possible to apply linear algebra operations to summarize and transform the dataset.